File: app/src/main/java/it/ncorti/emgvisualizer/myo/MyoSensor.java
/* This file is part of EmgVisualizer.

    EmgVisualizer is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    EmgVisualizer is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with EmgVisualizer.  If not, see <http://www.gnu.org/licenses/>.
*/
package it.ncorti.emgvisualizer.myo;

import android.app.Activity;
import android.bluetooth.BluetoothDevice;
import android.bluetooth.BluetoothGatt;
import android.bluetooth.BluetoothManager;
import android.bluetooth.le.BluetoothLeScanner;
import android.bluetooth.le.ScanCallback;
import android.bluetooth.le.ScanResult;
import android.content.Context;
import android.os.Handler;
import android.util.Log;

import it.ncorti.emgvisualizer.model.EventBusProvider;
import it.ncorti.emgvisualizer.model.RawDataPoint;
import it.ncorti.emgvisualizer.model.Sensor;
import it.ncorti.emgvisualizer.model.SensorConnectEvent;
import it.ncorti.emgvisualizer.model.SensorMeasuringEvent;
import it.ncorti.emgvisualizer.model.SensorRangeEvent;
import it.ncorti.emgvisualizer.model.SensorUpdateEvent;

/**
 * Class that represent the Myo EMG device sensor
 * @author Nicola
 */
public class MyoSensor extends Sensor {

    /**
     * TAG for debugging purpose
     */
    private static final String TAG = "MyoSensor";
    /**
     * Accelerometer sensor name
     */
    private static final String SENSOR_NAME = "Myo EMG Sensor";
    /**
     * Myo emg actual min value
     */
    private static final float MYO_MIN_VALUE = -120;
    /**
     * Myo emg actual max value
     */
    private static final float MYO_MAX_VALUE = 120;

    /**
     * Reference to bluetooth scanner
     */
    private BluetoothLeScanner mBluetoothScanner;
    /**
     * Bluethoot characteristic
     */
    private BluetoothGatt mBluetoothGatt;
    /**
     * Reference to GATT callbacks
     */
    private MyoGattCallback mMyoCallback;
    /**
     * Myo available commands
     */
    private MyoCommandList commandList;
    /**
     * Handler for thread execution
     */
    private Handler mHandler;
    /**
     * Myo Device Name
     */
    private String myoDeviceName;
    /**
     * Myo Device Name
     */
    private String myoDeviceAddress;
    /**
     * Reference to Context
     */
    private Context context;
    /**
     * Reference to Scan Callback
     */
    private ScanMyoListCallback mCallback;

    /**
     * Max Scan period in ms
     */
    private static final long SCAN_PERIOD = 5000;

    /**
     * Flag to detect if device is streaming or not
     */
    private boolean measuring = false;

    /**
     * Public constructor that requires
     * @param name      Myo Device name
     * @param hwAddress Myo Bluethoot address
     * @param context   Application context
     */
    public MyoSensor(String name, String hwAddress, Context context) {
        super(name + " - " + hwAddress, MYO_MIN_VALUE, MYO_MAX_VALUE);

        BluetoothManager mBluetoothManager = (BluetoothManager) context.getSystemService(Activity.BLUETOOTH_SERVICE);
        mBluetoothScanner = mBluetoothManager.getAdapter().getBluetoothLeScanner();

        mCallback = new ScanMyoListCallback();
        mHandler = new Handler();
        this.myoDeviceName = name;
        this.myoDeviceAddress = hwAddress;
        this.context = context;
        this.commandList = new MyoCommandList();
    }

    @Override
    public void startConnection() {
        mHandler.postDelayed(new Runnable() {
            @Override
            public void run() {
                mBluetoothScanner.stopScan(mCallback);
                EventBusProvider.postOnMainThread(new SensorConnectEvent(MyoSensor.this, false));
            }
        }, SCAN_PERIOD);
        mBluetoothScanner.startScan(mCallback);
    }

    @Override
    public void stopConnection() {
        closeBLEGatt();
    }

    @Override
    public boolean isConnected() {
        return (mBluetoothGatt != null);
    }

    @Override
    public void startMeasurement() {
        if (mBluetoothGatt == null ||
                !mMyoCallback.setMyoControlCommand(commandList.sendEmgOnly())) {
            Log.d(TAG, "Error start measurment");
            EventBusProvider.postOnMainThread(new SensorMeasuringEvent(this, false));
        } else {
            measuring = true;
            mMyoCallback.setMyoControlCommand(commandList.sendVibration3());
            EventBusProvider.postOnMainThread(new SensorMeasuringEvent(this, true));
        }
    }

    @Override
    public void stopMeasurement() {
        if (mBluetoothGatt == null ||
                !mMyoCallback.setMyoControlCommand(commandList.sendUnsetData())) {
            Log.d(TAG, "Error pause measurment");
            EventBusProvider.postOnMainThread(new SensorMeasuringEvent(this, true));
        } else {
            measuring = false;
            EventBusProvider.postOnMainThread(new SensorMeasuringEvent(this, false));
        }
    }

    @Override
    public boolean isMeasuring() {
        return measuring;
    }

    @Override
    public SensorUpdateEvent getUpdateEvent(RawDataPoint point) {
        return new MyoSensorUpdateEvent(this, point);
    }

    @Override
    public SensorRangeEvent getRangeEvent(float minValue, float maxValue) {
        return new MyoSensorRangeEvent(this, minValue, maxValue);
    }

    @Override
    public int getChannels() {
        return 8;
    }

    protected class ScanMyoListCallback extends ScanCallback {
        @Override
        public void onScanResult(int callbackType, ScanResult result) {
            super.onScanResult(callbackType, result);
            BluetoothDevice device = result.getDevice();
            if (myoDeviceName.equals(device.getName())) {
                mBluetoothScanner.stopScan(mCallback);
                // Trying to connect GATT
                mMyoCallback = new MyoGattCallback(MyoSensor.this);
                mBluetoothGatt = device.connectGatt(context, false, mMyoCallback);
                mMyoCallback.setBluetoothGatt(mBluetoothGatt);
                EventBusProvider.postOnMainThread(new SensorConnectEvent(MyoSensor.this, true));
            }
        }
    }

    /**
     * Private method for clossing GATT Profile connection
     */
    private void closeBLEGatt() {
        if (mBluetoothGatt == null) {
            return;
        }
        mMyoCallback.stopCallback();
        mBluetoothGatt.close();
        mBluetoothGatt = null;
    }
}

File: myonnaise/src/main/java/com/ncorti/myonnaise/MyoSensor.kt
package com.ncorti.myonnaise

import android.app.Activity
import android.bluetooth.BluetoothGatt
import android.bluetooth.BluetoothManager
import android.bluetooth.le.BluetoothLeScanner
import android.bluetooth.le.ScanCallback
import android.bluetooth.le.ScanResult
import android.content.Context
import android.os.Handler
import android.util.Log

import java.util.LinkedList

/**
 * Class that represent the Myo EMG device sensor
 */
class MyoSensor
/**
 * Public constructor that requires
 *
 * @param name      Myo Device name
 * @param hwAddress Myo Bluethoot address
 * @param context   Application context
 */
(
        /**
         * Myo Device Name
         */
        private val myoDeviceName: String,
        /**
         * Myo Device Name
         */
        private val myoDeviceAddress: String,
        /**
         * Reference to Context
         */
        private val context: Context) {

    /**
     * Reference to bluetooth scanner
     */
    private val mBluetoothScanner: BluetoothLeScanner
    /**
     * Bluethoot characteristic
     */
    private var mBluetoothGatt: BluetoothGatt? = null
    /**
     * Reference to GATT callbacks
     */
    private var mMyoCallback: OldMyoGattCallback? = null
    /**
     * Myo available commands
     */
    private val commandList: MyoCommandList
    /**
     * Handler for thread execution
     */
    private val mHandler: Handler
    /**
     * Reference to Scan Callback
     */
    private val mCallback: ScanMyoListCallback

    /**
     * Flag to detect if device is streaming or not
     */
    var isMeasuring = false
        private set

    val isConnected: Boolean
        get() = mBluetoothGatt != null

    val channels: Int
        get() = 8
    /**
     * Data point list
     */
    private val dataPoints = LinkedList<RawDataPoint>()
    /**
     * Max sensor value
     */
    /**
     * Getter for sensor min value
     *
     * @return Sensor min value
     */
    var minValue = Integer.MAX_VALUE.toFloat()
        private set
    /**
     * Min sensor value
     */
    /**
     * Getter for sensor max value
     *
     * @return Sensor max value
     */
    var maxValue = Integer.MIN_VALUE.toFloat()
        private set

    /**
     * Sensor name
     */
    /**
     * Getter for sensor name
     *
     * @return Sensor name
     */
    val name: String? = null

    /**
     * Method for getting short sensor name, used for Arff file (it's short name without spaces)
     *
     * @return Sensor short name
     */
    val shortName: String
        get() {
            var shortName = name!!.substring(0, 5).toLowerCase().replace(" ", "_")
            for (i in shortName.length..4)
                shortName += "_"
            return shortName
        }

    /**
     * Getter for sensor status string (for displaying purposes
     *
     * @return Sensor status string
     */
    //status += "Received RAW points: <b>" + getDataPoints().size() + "</b></p>";
    val statusString: String
        get() {
            var status = "<p>Device is currenty: "
            if (isConnected) {
                status += " <b>Connected.</b><br>"
            } else {
                status += " <b>Disconnected.</b><br>"
            }
            if (isMeasuring)
                status += " Raw data receiving is: <b>Active.</b><br>"
            else
                status += " Raw data receiving is: <b>Inactive.</b><br>"
            status += "</p>"
            return status
        }

    init {

        val mBluetoothManager = context.getSystemService(Activity.BLUETOOTH_SERVICE) as BluetoothManager
        mBluetoothScanner = mBluetoothManager.adapter.bluetoothLeScanner

        mCallback = ScanMyoListCallback()
        mHandler = Handler()
        this.commandList = MyoCommandList()
    }

    fun startConnection() {
        mHandler.postDelayed({
            mBluetoothScanner.stopScan(mCallback)
            //                EventBusProvider.postOnMainThread(new SensorConnectEvent(MyoSensor.this, false));
        }, SCAN_PERIOD)
        mBluetoothScanner.startScan(mCallback)
    }

    fun stopConnection() {
        closeBLEGatt()
    }

    fun startMeasurement() {
        if (mBluetoothGatt == null || !mMyoCallback!!.setMyoControlCommand(commandList.sendEmgOnly())) {
            Log.d(TAG, "Error start measurment")
        } else {
            isMeasuring = true
            mMyoCallback!!.setMyoControlCommand(commandList.sendVibration3())
        }
    }

    fun stopMeasurement() {
        if (mBluetoothGatt == null || !mMyoCallback!!.setMyoControlCommand(commandList.sendUnsetData())) {
            Log.d(TAG, "Error pause measurment")
        } else {
            isMeasuring = false
        }
    }

    protected inner class ScanMyoListCallback : ScanCallback() {
        override fun onScanResult(callbackType: Int, result: ScanResult) {
            super.onScanResult(callbackType, result)
            val device = result.device
            if (myoDeviceName == device.name) {
                mBluetoothScanner.stopScan(mCallback)
                // Trying to connect GATT
                mMyoCallback = OldMyoGattCallback(this@MyoSensor)
                mBluetoothGatt = device.connectGatt(context, false, mMyoCallback)
                mMyoCallback!!.setBluetoothGatt(mBluetoothGatt)
                //                EventBusProvider.postOnMainThread(new SensorConnectEvent(MyoSensor.this, true));
            }
        }
    }

    /**
     * Private method for clossing GATT Profile connection
     */
    private fun closeBLEGatt() {
        if (mBluetoothGatt == null) {
            return
        }
        mMyoCallback!!.stopCallback()
        mBluetoothGatt!!.close()
        mBluetoothGatt = null
    }

    /**
     * Getter for data point list.
     * The list is cloned before returning.
     *
     * @return Data point list
     */
    @Synchronized
    fun getDataPoints(): LinkedList<RawDataPoint> {
        return dataPoints.clone() as LinkedList<RawDataPoint>
    }

    /**
     * Method for adding a point to the sensor data point list
     *
     * @param dataPoint New point to be added
     */
    @Synchronized
    fun addDataPoint(dataPoint: RawDataPoint) {
        dataPoints.addLast(dataPoint)

        if (dataPoints.size > MAX_DATA_POINTS) {
            dataPoints.removeFirst()
        }

        // Check for new range event
        var newLimits = false

        for (value in dataPoint.values) {
            if (value > maxValue) {
                maxValue = value
                newLimits = true
            }
            if (value < minValue) {
                minValue = value
                newLimits = true
            }
        }

        if (newLimits) {
            Log.d(TAG, "New range for sensor $name: $minValue - $maxValue")
            //            EventBusProvider.postOnMainThread(getRangeEvent(minValue, maxValue));
        }
        //        SensorUpdateEvent updateEvent = getUpdateEvent(dataPoint);
        //        EventBusProvider.postOnMainThread(updateEvent);
    }

    /**
     * Method for clearing data point list
     */
    @Synchronized
    fun clearDataPoints() {
        dataPoints.clear()
    }

    companion object {

        /**
         * TAG for debugging purpose
         */
        private val TAG = "MyoSensor"
        /**
         * Accelerometer sensor name
         */
        private val SENSOR_NAME = "Myo EMG Sensor"
        /**
         * Myo emg actual min value
         */
        private val MYO_MIN_VALUE = -120f
        /**
         * Myo emg actual max value
         */
        private val MYO_MAX_VALUE = 120f

        /**
         * Max Scan period in ms
         */
        private val SCAN_PERIOD: Long = 5000

        /**
         * Data point list size
         */
        private val MAX_DATA_POINTS = 5000
    }
}
